pkg/backend/display/rows.go

```diff
@@ -305,17 +305,41 @@ func (data *resourceRowData) ColorizedColumns() []string {
 
 	diagInfo := data.diagInfo
 
-	if done {
-		failed := data.failed || diagInfo.ErrorCount > 0
-		columns[statusColumn] = data.display.getStepDoneDescription(step, failed)
-	} else {
-		columns[statusColumn] = data.display.getStepInProgressDescription(step)
-	}
+	failed := data.failed || diagInfo.ErrorCount > 0
 
+	columns[statusColumn] = data.display.getStepStatus(step, done, failed)
 	columns[infoColumn] = data.getInfoColumn()
 	return columns
 }
 
+// addRetainStatusFlag adds a "[retain]" suffix to the input string if the resource is marked as
+// RetainOnDelete and the step will discard the resource.
+func addRetainStatusFlag(status string, step engine.StepEventMetadata) string {
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old == nil {
+		return status
+	}
+
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old.State == nil {
+		return status
+	}
+
+	// If a resource is not marked as RetainOnDelete, return the status string as is.
+	if !step.Old.State.RetainOnDelete {
+		return status
+	}
+	switch step.Op {
+	// Deletes and Replacements should indicate retain on delete behavior as they can leave
+	// untracked resources in the environment.
```

Comment:
e.g. this one is a useful comment because it's adding some more context that may not be obvious from the name and documentation of this function and surrounding code.